llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp

```diff
@@ -3357,7 +3362,9 @@ Value *LibCallSimplifier::optimizePrintFString(CallInst *CI, IRBuilderBase &B) {
     // Create a string literal with no \n on it.  We expect the constant merge
     // pass to be run after this pass, to merge duplicate strings.
     FormatStr = FormatStr.drop_back();
-    Value *GV = B.CreateGlobalString(FormatStr, "str");
+    Value *GV = B.CreateGlobalString(
+        FormatStr, "str",
+        CI->getArgOperand(0)->getType()->getPointerAddressSpace());
```

Comment:
If you're going to touch CreateGlobalString, please fix all the calls in this file at the same time.

On targets with multiple address-spaces, is it actually legal to create a global like this?  I thought the usual convention was to create a global in some specific address-space, then addrspacecast it to the correct address-space.

Reply:
Generally should try to use a context address space. For a newly synthesized global, there's DL.getDefaultGlobalsAddressSpace

Reply:
In this specific line we're replacing one global string with another, so it makes sense to use the original address space. I'll look into other uses within this file.

Reply:
getConstantStringInfo looks through addrspacecasts, so `CI->getArgOperand(0)->getType()->getPointerAddressSpace()` isn't necessarily the "original" address-space.

Reply:
Alright, let's just use the default globals address space. I'm going to split this component into a separate PR.

Reply:
Split off into https://github.com/llvm/llvm-project/pull/118729

Writing the test for this exposed more places where we were failing to propagate address spaces correctly, which are addressed in that PR as well.